synadm/cli/__init__.py

```diff
@@ -175,6 +183,48 @@ def output(self, data):
         """
         click.echo(self.formatter(data))
 
+    def retrieve_homeserver_name(self, uri):
+        """Try to retrieve the homeserver name.
+
+        When homeserver is set in the config already, it's just returned and
+        nothing is tried to be fetched automatically. If not, either the
+        location of the federation API is looked up via a .well-known resource
+        or a DNS SRV lookup. This depends on the server_discovery setting in the
+        config.
+
+        Args:
+            uri (string): proto://name:port or proto://fqdn:port
+
+        Returns:
+            string: hostname, FQDN or DOMAIN; or None on errors.
+        """
+        if self.config["homeserver"] != "auto-retrieval":
+            return  self.config["homeserver"]
+
+        if self.config["server_discovery"] == "well-known":
+            federation_uri = self.misc_request.federation_uri_well_known(uri)
+            if not federation_uri:
+                return None
+            return self.matrix_api.server_name_keys_api(federation_uri)
+        elif self.config["server_discovery"] == "dns":
```

Comment:
@Ascurius please review and test. I tried to add some error handling to the DNS SRV resolving. Also tried to make use of the possibilities (attributes target, port) of the dns.resolver object in the else branch of the try/except clause. No extra sanity check of whether a record 0 is there. Do you think that is sufficient? Please do some testing with bollocks values for hostname and simulated network errors.

Reply:
Sry for my late response. I have reviewed and tested your changes. As far as I can see, everything works as intended, I could not find any bugs 👍

Reply:
Ah one thing to look closely: I might have forgotten to document newly implemented args in docstrings (or adapt to changed behaviour) Please doublecheck! Thanks a ton!!